File: parsing/action_filter.go
package parsing

import (
	"regexp"
	"strings"
	"unicode"
)

type actionMode struct{ e uint }

var (
	notStarted    = actionMode{0}
	toolCallID    = actionMode{1}
	toolCallIDEnd = actionMode{2}
	toolName      = actionMode{3}
	toolNameEnd   = actionMode{4}
	paramName     = actionMode{5}
	paramValue    = actionMode{6}
	toolEnd       = actionMode{7}
	paramNameEnd  = actionMode{8}
	paramValueEnd = actionMode{9}
	rawParam      = actionMode{10}
)

type filterAction struct {
	mode             actionMode
	curToolCallIndex int
	trimLeft         bool

	// Parameter metadata
	curParamName     string
	curParamState    paramState
	paramValueBuffer string
}

var (
	toolCallIDRegex = regexp.MustCompile(`"tool_call_id":\s*"`)
	toolNameRegex   = regexp.MustCompile(`"tool_name":\s*"`)
	paramRegex      = regexp.MustCompile(`"parameters":\s*{\s*"`)
	rawParamRegex   = regexp.MustCompile(`"parameters":\s*`)
	paramNameRegex  = regexp.MustCompile(`\s*:\s*`)

	llamaToolNameRegex = regexp.MustCompile(`"name":\s*"`)
)

func (f *filter) ParseActions(str string) ([]FilterOutput, int) {
	// If the last character is \ then we are waiting for more - incase of an escaped character
	if str == "" || str[len(str)-1] == '\\' {
		return nil, 0
	}
	switch f.actionMetaData.mode {
	case notStarted, toolEnd:
		return f.HandleBeforeTool(str, f.hasToolCallID)
	case toolCallID:
		return f.HandleInToolCallID(str)
	case toolCallIDEnd:
		return f.HandleToolCallIDEnd(str)
	case toolName:
		return f.HandleInToolName(str)
	case toolNameEnd:
		return f.HandleToolNameEnd(str)
	case rawParam:
		return f.HandleRawParam(str)
	case paramName:
		return f.HandleParamName(str)
	case paramNameEnd:
		return f.HandleEndOfParamName(str)
	case paramValue:
		return f.HandleParamValue(str)
	case paramValueEnd:
		return f.HandleParamValueEnd(str)
	}
	return nil, 0
}

// We are waiting for "tool_name": " (or tool_call_id) the regex allows all the whitespace in between
func (f *filter) HandleBeforeTool(str string, checkCallID bool) ([]FilterOutput, int) {
	var indices []int
	var mode actionMode

	switch {
	case f.llamaToolParsing:
		indices = llamaToolNameRegex.FindStringIndex(str)
		mode = toolName
	case checkCallID:
		indices = toolCallIDRegex.FindStringIndex(str)
		mode = toolCallID
	default:
		indices = toolNameRegex.FindStringIndex(str)
		mode = toolName
	}
	if indices == nil {
		return nil, 0
	}
	f.actionMetaData.mode = mode
	f.actionMetaData.trimLeft = true // Start with trimming left
	out, rem := f.ParseActions(str[indices[1]:])
	return out, rem + len(str[:indices[1]])
}

// We are expecting for the end of the tool call id with "
func (f *filter) HandleInToolCallID(str string) ([]FilterOutput, int) {
	idx := findNonEscapedChar(str, '"')
	if idx == -1 {
		// Wait till end of quotes
		return nil, 0
	}
	out := f.sendToolCallIDChunk(str[:idx])
	f.actionMetaData.mode = toolCallIDEnd
	o, r := f.ParseActions(str[idx:]) // one for the quote
	return append(out, o...), r + len(str[:idx]) + 1
}

func (f *filter) HandleToolCallIDEnd(str string) ([]FilterOutput, int) {
	// Looking for the name next
	return f.HandleBeforeTool(str, false)
}

// We are expecting for the end of the tool name with "
func (f *filter) HandleInToolName(str string) ([]FilterOutput, int) {
	idx := findNonEscapedChar(str, '"')
	if idx == -1 {
		// Wait till end of tool name
		return nil, 0
	}
	out := f.sendToolNameChunk(str[:idx])
	f.actionMetaData.mode = toolNameEnd
	o, r := f.ParseActions(str[idx:])
	return append(out, o...), r + len(str[:idx]) + 1 // one for the quote
}

// We are waiting for "parameters": { " the regex allows all the whitespace in between
func (f *filter) HandleToolNameEnd(str string) ([]FilterOutput, int) {
	indices := paramRegex.FindStringIndex(str)
	if indices == nil {
		idx := strings.Index(str, `}`) // no parameters (directly answer for example)
		if idx == -1 {
			return nil, 0
		}
		f.actionMetaData.mode = toolEnd
		f.actionMetaData.curToolCallIndex++
		f.actionMetaData.curParamName = ""
		out, rem := f.ParseActions(str[idx:])
		return out, rem + len(str[:idx])
	}
	if f.streamProcessedParams {
		f.actionMetaData.mode = paramName
		out, rem := f.ParseActions(str[indices[1]:])
		return out, rem + len(str[:indices[1]])
	}
	f.actionMetaData.mode = rawParam
	// Without {
	indices = rawParamRegex.FindStringIndex(str)
	out, rem := f.ParseActions(str[indices[1]:])
	return out, rem + len(str[:indices[1]])
}

// When handling the parameters raw we just send the whole string
// Waiting for a valid JSON parameter to know we are at the end of the param
func (f *filter) HandleRawParam(str string) ([]FilterOutput, int) {
	idx := findValidJSONValue(f.actionMetaData.paramValueBuffer, str)
	if idx == -1 {
		// If we don't find valid json then return the whole string and wait for more
		out := f.sendRawParamChunkWithoutIndentation(str)
		f.actionMetaData.paramValueBuffer += str
		return out, len(str)
	}

	// We have a valid JSON value
	out := f.sendRawParamChunkWithoutIndentation(str[:idx])
	f.actionMetaData.paramValueBuffer = ""
	f.actionMetaData.curToolCallIndex++
	f.actionMetaData.mode = toolEnd
	o, r := f.ParseActions(str[idx:])
	return append(out, o...), r + len(str[:idx])
}

// NumSpaceToRemovePerLine is the number of spaces to remove from each line when streaming raw parameters.
// (excluding the first line which is just "{\n"). This is exactly two levels of indentation.
const NumSpaceToRemovePerLine = 8

func (f *filter) sendRawParamChunkWithoutIndentation(str string) []FilterOutput {
	trimmedStrBldr := strings.Builder{}
	for _, c := range str {
		switch {
		case c == '\n':
			f.rawParamIndentLengthRemoved = 0
			f.sawNonWhitespaceInCurrentline = false
		case unicode.IsSpace(c):
			if f.rawParamIndentLengthRemoved < NumSpaceToRemovePerLine && !f.sawNonWhitespaceInCurrentline {
				f.rawParamIndentLengthRemoved++
				continue
			}
		default:
			f.sawNonWhitespaceInCurrentline = true
		}
		trimmedStrBldr.WriteRune(c)
	}
	return f.sendRawParamChunk(trimmedStrBldr.String())
}

// We are expecting for the end of the param name with "
func (f *filter) HandleParamName(str string) ([]FilterOutput, int) {
	idx := findNonEscapedChar(str, '"')
	if idx == -1 {
		// Wait till end of parameter name
		return nil, 0
	}
	out := f.sendParamNameChunk(str[:idx])
	f.actionMetaData.mode = paramNameEnd
	o, r := f.ParseActions(str[idx:])
	return append(out, o...), r + len(str[:idx]) + 1 // one for the quote
}

// We are expecting : for the start of the value the regex allows for whitespace in between
func (f *filter) HandleEndOfParamName(str string) ([]FilterOutput, int) {
	indices := paramNameRegex.FindStringIndex(str)
	if indices == nil {
		return nil, 0
	}
	f.actionMetaData.mode = paramValue
	out, rem := f.ParseActions(str[indices[1]:])
	return out, rem + len(str[:indices[1]])
}

// We are expecting for the start of the param name with " this is only when we have a next param
func (f *filter) HandleParamValueEnd(str string) ([]FilterOutput, int) {
	idx := strings.Index(str, `"`)
	if idx == -1 {
		return nil, 0
	}
	f.actionMetaData.mode = paramName
	out, rem := f.ParseActions(str[idx+1:])
	return out, rem + len(str[:idx]) + 1
}

func (f *filter) sendToolCallIDChunk(str string) []FilterOutput {
	if str == "" || !f.streamToolActions {
		return nil
	}
	return []FilterOutput{{
		ToolCalls: &FilterToolCallDelta{
			Index: f.actionMetaData.curToolCallIndex,
			ID:    str,
		},
	}}
}

func (f *filter) sendToolNameChunk(str string) []FilterOutput {
	if str == "" || !f.streamToolActions {
		return nil
	}
	return []FilterOutput{{
		ToolCalls: &FilterToolCallDelta{
			Index: f.actionMetaData.curToolCallIndex,
			Name:  str,
		},
	}}
}

func (f *filter) sendParamNameChunk(str string) []FilterOutput {
	if str == "" || !f.streamToolActions {
		return nil
	}
	f.actionMetaData.curParamName = str
	return []FilterOutput{{
		ToolCalls: &FilterToolCallDelta{
			Index: f.actionMetaData.curToolCallIndex,
			ParamDelta: &FilterToolParameter{
				Name: str,
			},
		},
	}}
}

func (f *filter) sendRawParamChunk(str string) []FilterOutput {
	if str == "" || !f.streamToolActions {
		return nil
	}
	return []FilterOutput{{
		ToolCalls: &FilterToolCallDelta{
			Index:         f.actionMetaData.curToolCallIndex,
			RawParamDelta: str,
		},
	}}
}

func (f *filter) sendParamValueChunk(str string) ([]FilterOutput, int) {
	// Just for parameter value we need to trim the right whitespace as the value doesn't have a defined end like "
	trimmedStr := strings.TrimRightFunc(str, unicode.IsSpace)
	// Trim left just at the start
	if f.actionMetaData.trimLeft {
		trimmedStr = strings.TrimLeftFunc(trimmedStr, unicode.IsSpace)
	}
	if trimmedStr == "" || !f.streamToolActions {
		return nil, 0
	}
	// We have seen a value so stop trimming left
	f.actionMetaData.trimLeft = false
	return []FilterOutput{{
		ToolCalls: &FilterToolCallDelta{
			Index: f.actionMetaData.curToolCallIndex,
			ParamDelta: &FilterToolParameter{
				Name:       f.actionMetaData.curParamName,
				ValueDelta: trimmedStr,
			},
		},
	}}, len(str)
}

func findNonEscapedChar(str string, char byte) int {
	// there needs to be an even number of \ before the character
	// if there is an odd number then the character is escaped
	for i := 0; i < len(str); i++ {
		if str[i] == char {
			// check if the character is escaped
			escaped := false
			for j := i - 1; j >= 0; j-- {
				if str[j] != '\\' {
					break
				}
				escaped = !escaped
			}
			if !escaped {
				return i
			}
		}
	}
	return -1
}

File: rust/src/action_filter.rs
use crate::filter::FilterImpl;
use crate::param_filter::ParamState;
use crate::types::*;
use regex::Regex;

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub(crate) enum ActionMode {
    NotStarted,
    ToolCallID,
    ToolCallIDEnd,
    ToolName,
    ToolNameEnd,
    ParamName,
    ParamValue,
    ToolEnd,
    ParamNameEnd,
    ParamValueEnd,
    RawParam,
}

#[derive(Debug, Clone)]
pub(crate) struct FilterAction {
    pub mode: ActionMode,
    pub cur_tool_call_index: usize,
    pub trim_left: bool,
    pub cur_param_name: String,
    pub cur_param_state: ParamState,
    pub param_value_buffer: String,
}

impl FilterAction {
    pub fn new() -> Self {
        Self {
            mode: ActionMode::NotStarted,
            cur_tool_call_index: 0,
            trim_left: false,
            cur_param_name: String::new(),
            cur_param_state: ParamState::Beginning,
            param_value_buffer: String::new(),
        }
    }
}

impl FilterImpl {
    pub(crate) fn parse_actions(&mut self, s: &str) -> (Vec<FilterOutput>, usize) {
        if s.is_empty() || s.ends_with('\\') {
            return (Vec::new(), 0);
        }

        match self.action_metadata.mode {
            ActionMode::NotStarted | ActionMode::ToolEnd => {
                self.handle_before_tool(s, self.has_tool_call_id)
            }
            ActionMode::ToolCallID => self.handle_in_tool_call_id(s),
            ActionMode::ToolCallIDEnd => self.handle_tool_call_id_end(s),
            ActionMode::ToolName => self.handle_in_tool_name(s),
            ActionMode::ToolNameEnd => self.handle_tool_name_end(s),
            ActionMode::RawParam => self.handle_raw_param(s),
            ActionMode::ParamName => self.handle_param_name(s),
            ActionMode::ParamNameEnd => self.handle_end_of_param_name(s),
            ActionMode::ParamValue => self.handle_param_value(s),
            ActionMode::ParamValueEnd => self.handle_param_value_end(s),
        }
    }

    fn handle_before_tool(&mut self, s: &str, check_call_id: bool) -> (Vec<FilterOutput>, usize) {
        let (regex, mode) = if self.llama_tool_parsing {
            (Regex::new(r#""name":\s*""#).unwrap(), ActionMode::ToolName)
        } else if check_call_id {
            (
                Regex::new(r#""tool_call_id":\s*""#).unwrap(),
                ActionMode::ToolCallID,
            )
        } else {
            (
                Regex::new(r#""tool_name":\s*""#).unwrap(),
                ActionMode::ToolName,
            )
        };

        if let Some(mat) = regex.find(s) {
            self.action_metadata.mode = mode;
            self.action_metadata.trim_left = true;
            let (out, rem) = self.parse_actions(&s[mat.end()..]);
            (out, rem + mat.end())
        } else {
            (Vec::new(), 0)
        }
    }

    fn handle_in_tool_call_id(&mut self, s: &str) -> (Vec<FilterOutput>, usize) {
        if let Some(idx) = find_non_escaped_char(s, '"') {
            let out = self.send_tool_call_id_chunk(&s[..idx]);
            self.action_metadata.mode = ActionMode::ToolCallIDEnd;
            let (o, r) = self.parse_actions(&s[idx..]);
            let mut result = out;
            result.extend(o);
            (result, r + idx + 1)
        } else {
            (Vec::new(), 0)
        }
    }

    fn handle_tool_call_id_end(&mut self, s: &str) -> (Vec<FilterOutput>, usize) {
        self.handle_before_tool(s, false)
    }

    fn handle_in_tool_name(&mut self, s: &str) -> (Vec<FilterOutput>, usize) {
        if let Some(idx) = find_non_escaped_char(s, '"') {
            let out = self.send_tool_name_chunk(&s[..idx]);
            self.action_metadata.mode = ActionMode::ToolNameEnd;
            let (o, r) = self.parse_actions(&s[idx..]);
            let mut result = out;
            result.extend(o);
            (result, r + idx + 1)
        } else {
            (Vec::new(), 0)
        }
    }

    fn handle_tool_name_end(&mut self, s: &str) -> (Vec<FilterOutput>, usize) {
        let param_regex = Regex::new(r#""parameters":\s*\{\s*""#).unwrap();

        if let Some(mat) = param_regex.find(s) {
            if self.stream_processed_params {
                self.action_metadata.mode = ActionMode::ParamName;
                let (out, rem) = self.parse_actions(&s[mat.end()..]);
                return (out, rem + mat.end());
            }
            self.action_metadata.mode = ActionMode::RawParam;
            let raw_param_regex = Regex::new(r#""parameters":\s*"#).unwrap();
            if let Some(mat) = raw_param_regex.find(s) {
                let (out, rem) = self.parse_actions(&s[mat.end()..]);
                return (out, rem + mat.end());
            }
        }

        if let Some(idx) = s.find('}') {
            self.action_metadata.mode = ActionMode::ToolEnd;
            self.action_metadata.cur_tool_call_index += 1;
            self.action_metadata.cur_param_name.clear();
            let (out, rem) = self.parse_actions(&s[idx..]);
            (out, rem + idx)
        } else {
            (Vec::new(), 0)
        }
    }

    fn handle_raw_param(&mut self, s: &str) -> (Vec<FilterOutput>, usize) {
        use crate::param_filter::find_valid_json_value;

        let idx = find_valid_json_value(&self.action_metadata.param_value_buffer, s);

        if idx == usize::MAX {
            let out = self.send_raw_param_chunk_without_indentation(s);
            self.action_metadata.param_value_buffer.push_str(s);
            (out, s.len())
        } else {
            let out = self.send_raw_param_chunk_without_indentation(&s[..idx]);
            self.action_metadata.param_value_buffer.clear();
            self.action_metadata.cur_tool_call_index += 1;
            self.action_metadata.mode = ActionMode::ToolEnd;
            let (o, r) = self.parse_actions(&s[idx..]);
            let mut result = out;
            result.extend(o);
            (result, r + idx)
        }
    }

    const NUM_SPACE_TO_REMOVE_PER_LINE: usize = 8;

    fn send_raw_param_chunk_without_indentation(&mut self, s: &str) -> Vec<FilterOutput> {
        let mut trimmed_str = String::new();

        for c in s.chars() {
            match c {
                '\n' => {
                    self.raw_param_indent_length_removed = 0;
                    self.saw_non_whitespace_in_current_line = false;
                    trimmed_str.push(c);
                }
                c if c.is_whitespace() => {
                    if self.raw_param_indent_length_removed < Self::NUM_SPACE_TO_REMOVE_PER_LINE
                        && !self.saw_non_whitespace_in_current_line
                    {
                        self.raw_param_indent_length_removed += 1;
                        continue;
                    }
                    trimmed_str.push(c);
                }
                _ => {
                    self.saw_non_whitespace_in_current_line = true;
                    trimmed_str.push(c);
                }
            }
        }

        self.send_raw_param_chunk(&trimmed_str)
    }

    fn handle_param_name(&mut self, s: &str) -> (Vec<FilterOutput>, usize) {
        if let Some(idx) = find_non_escaped_char(s, '"') {
            let out = self.send_param_name_chunk(&s[..idx]);
            self.action_metadata.mode = ActionMode::ParamNameEnd;
            let (o, r) = self.parse_actions(&s[idx..]);
            let mut result = out;
            result.extend(o);
            (result, r + idx + 1)
        } else {
            (Vec::new(), 0)
        }
    }

    fn handle_end_of_param_name(&mut self, s: &str) -> (Vec<FilterOutput>, usize) {
        let param_name_regex = Regex::new(r#"\s*:\s*"#).unwrap();

        if let Some(mat) = param_name_regex.find(s) {
            self.action_metadata.mode = ActionMode::ParamValue;
            let (out, rem) = self.parse_actions(&s[mat.end()..]);
            (out, rem + mat.end())
        } else {
            (Vec::new(), 0)
        }
    }

    fn handle_param_value_end(&mut self, s: &str) -> (Vec<FilterOutput>, usize) {
        if let Some(idx) = s.find('"') {
            self.action_metadata.mode = ActionMode::ParamName;
            let (out, rem) = self.parse_actions(&s[idx + 1..]);
            (out, rem + idx + 1)
        } else {
            (Vec::new(), 0)
        }
    }

    fn send_tool_call_id_chunk(&self, s: &str) -> Vec<FilterOutput> {
        if s.is_empty() || !self.stream_tool_actions {
            return Vec::new();
        }

        vec![FilterOutput {
            tool_calls: Some(FilterToolCallDelta {
                index: self.action_metadata.cur_tool_call_index,
                id: s.to_string(),
                ..Default::default()
            }),
            ..Default::default()
        }]
    }

    fn send_tool_name_chunk(&self, s: &str) -> Vec<FilterOutput> {
        if s.is_empty() || !self.stream_tool_actions {
            return Vec::new();
        }

        vec![FilterOutput {
            tool_calls: Some(FilterToolCallDelta {
                index: self.action_metadata.cur_tool_call_index,
                name: s.to_string(),
                ..Default::default()
            }),
            ..Default::default()
        }]
    }

    fn send_param_name_chunk(&mut self, s: &str) -> Vec<FilterOutput> {
        if s.is_empty() || !self.stream_tool_actions {
            return Vec::new();
        }

        self.action_metadata.cur_param_name = s.to_string();

        vec![FilterOutput {
            tool_calls: Some(FilterToolCallDelta {
                index: self.action_metadata.cur_tool_call_index,
                param_delta: Some(FilterToolParameter {
                    name: s.to_string(),
                    value_delta: String::new(),
                }),
                ..Default::default()
            }),
            ..Default::default()
        }]
    }

    fn send_raw_param_chunk(&self, s: &str) -> Vec<FilterOutput> {
        if s.is_empty() || !self.stream_tool_actions {
            return Vec::new();
        }

        vec![FilterOutput {
            tool_calls: Some(FilterToolCallDelta {
                index: self.action_metadata.cur_tool_call_index,
                raw_param_delta: s.to_string(),
                ..Default::default()
            }),
            ..Default::default()
        }]
    }

    pub(crate) fn send_param_value_chunk(&mut self, s: &str) -> (Vec<FilterOutput>, usize) {
        let trimmed_str = s.trim_end();
        let trimmed_str = if self.action_metadata.trim_left {
            trimmed_str.trim_start()
        } else {
            trimmed_str
        };

        if trimmed_str.is_empty() || !self.stream_tool_actions {
            return (Vec::new(), 0);
        }

        self.action_metadata.trim_left = false;

        (
            vec![FilterOutput {
                tool_calls: Some(FilterToolCallDelta {
                    index: self.action_metadata.cur_tool_call_index,
                    param_delta: Some(FilterToolParameter {
                        name: self.action_metadata.cur_param_name.clone(),
                        value_delta: trimmed_str.to_string(),
                    }),
                    ..Default::default()
                }),
                ..Default::default()
            }],
            s.len(),
        )
    }
}

fn find_non_escaped_char(s: &str, ch: char) -> Option<usize> {
    let bytes = s.as_bytes();
    for i in 0..bytes.len() {
        if bytes[i] == ch as u8 {
            let mut escaped = false;
            for j in (0..i).rev() {
                if bytes[j] != b'\\' {
                    break;
                }
                escaped = !escaped;
            }
            if !escaped {
                return Some(i);
            }
        }
    }
    None
}
